// evergreen-ui/commonjs/table/src/TextTableHeaderCell.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _typography = require("../../typography");

var _TableHeaderCell = _interopRequireDefault(require("./TableHeaderCell"));

var TextTableHeaderCell =
/*#__PURE__*/
function (_PureComponent) {
  (0, _inherits2.default)(TextTableHeaderCell, _PureComponent);

  function TextTableHeaderCell() {
    (0, _classCallCheck2.default)(this, TextTableHeaderCell);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(TextTableHeaderCell).apply(this, arguments));
  }

  (0, _createClass2.default)(TextTableHeaderCell, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          children = _this$props.children,
          textProps = _this$props.textProps,
          isSortable = _this$props.isSortable,
          sortOrder = _this$props.sortOrder,
          props = (0, _objectWithoutProperties2.default)(_this$props, ["children", "textProps", "isSortable", "sortOrder"]);
      return _react.default.createElement(_TableHeaderCell.default, props, _react.default.createElement(_typography.Text, (0, _extends2.default)({
        fontWeight: 500,
        size: 300,
        flex: "1"
      }, textProps), children, ' '));
    }
  }]);
  return TextTableHeaderCell;
}(_react.PureComponent);

exports.default = TextTableHeaderCell;
TextTableHeaderCell.displayName = "TextTableHeaderCell";
(0, _defineProperty2.default)(TextTableHeaderCell, "propTypes", (0, _objectSpread2.default)({}, _TableHeaderCell.default.propTypes, {
  /**
   * Pass additional props to the Text component.
   */
  textProps: _propTypes.default.objectOf(_propTypes.default.object)
}));
