// ui-box/lib/enhancers/background.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.propEnhancers = exports.propValidators = exports.propAliases = exports.propTypes = undefined;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _getCss = require('../get-css');

var _getCss2 = _interopRequireDefault(_getCss);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var propTypes = exports.propTypes = {
  background: _propTypes2.default.string,
  backgroundBlendMode: _propTypes2.default.string,
  backgroundClip: _propTypes2.default.string,
  backgroundColor: _propTypes2.default.string,
  backgroundImage: _propTypes2.default.string,
  backgroundOrigin: _propTypes2.default.string,
  backgroundPosition: _propTypes2.default.string,
  backgroundRepeat: _propTypes2.default.string,
  backgroundSize: _propTypes2.default.string
};

var propAliases = exports.propAliases = {};

var propValidators = exports.propValidators = {};

var _background = {
  className: 'bg',
  cssName: 'background',
  jsName: 'background',
  isPrefixed: true,
  complexValue: true
};
var _backgroundColor = {
  className: 'bg-clr',
  cssName: 'background-color',
  jsName: 'backgroundColor'
};
var _backgroundImage = {
  className: 'bg-img',
  cssName: 'background-image',
  jsName: 'backgroundImage',
  isPrefixed: true,
  complexValue: true
};
var _backgroundPosition = {
  className: 'bg-pos',
  cssName: 'background-position',
  jsName: 'backgroundPosition'
};
var _backgroundSize = {
  className: 'bg-siz',
  cssName: 'background-size',
  jsName: 'backgroundSize'
};
var _backgroundOrigin = {
  className: 'bg-orgn',
  cssName: 'background-origin',
  jsName: 'backgroundOrigin'
};
var _backgroundRepeat = {
  className: 'bg-rpt',
  cssName: 'background-repeat',
  jsName: 'backgroundRepeat'
};
var _backgroundClip = {
  className: 'bg-clp',
  cssName: 'background-clip',
  jsName: 'backgroundClip'
};
var _backgroundBlendMode = {
  className: 'bg-blnd-md',
  cssName: 'background-blend-mode',
  jsName: 'backgroundBlendMode'
};

var propEnhancers = exports.propEnhancers = {
  background: function background(value) {
    return (0, _getCss2.default)(_background, value);
  },
  backgroundBlendMode: function backgroundBlendMode(value) {
    return (0, _getCss2.default)(_backgroundBlendMode, value);
  },
  backgroundClip: function backgroundClip(value) {
    return (0, _getCss2.default)(_backgroundClip, value);
  },
  backgroundColor: function backgroundColor(value) {
    return (0, _getCss2.default)(_backgroundColor, value);
  },
  backgroundImage: function backgroundImage(value) {
    return (0, _getCss2.default)(_backgroundImage, value);
  },
  backgroundOrigin: function backgroundOrigin(value) {
    return (0, _getCss2.default)(_backgroundOrigin, value);
  },
  backgroundPosition: function backgroundPosition(value) {
    return (0, _getCss2.default)(_backgroundPosition, value);
  },
  backgroundRepeat: function backgroundRepeat(value) {
    return (0, _getCss2.default)(_backgroundRepeat, value);
  },
  backgroundSize: function backgroundSize(value) {
    return (0, _getCss2.default)(_backgroundSize, value);
  }
};
