// ui-box/lib/enhancers/border-radius.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.propEnhancers = exports.propValidators = exports.propAliases = exports.propTypes = undefined;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _getCss = require('../get-css');

var _getCss2 = _interopRequireDefault(_getCss);

var _regex = require('../utils/regex');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var propTypes = exports.propTypes = {
  borderBottomLeftRadius: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  borderBottomRightRadius: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  borderRadius: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  borderTopLeftRadius: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  borderTopRightRadius: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number])
};

var propAliases = exports.propAliases = {
  borderRadius: ['borderBottomLeftRadius', 'borderBottomRightRadius', 'borderTopLeftRadius', 'borderTopRightRadius']
};

var propValidators = exports.propValidators = {};

if (process.env.NODE_ENV !== 'production') {
  propValidators.borderRadius = function (value) {
    if (_regex.spacesOutsideParentheses.test(value)) {
      return 'multiple values (\u201C' + value + '\u201D) aren\u055At supported with \u201CborderRadius\u201D. Use \u201CborderBottomLeftRadius\u201D, \u201CborderBottomRightRadius\u201D \u201CborderTopLeftRadius\u201D and \u201CborderTopRightRadius\u201D instead.';
    }
  };
}

var _borderTopLeftRadius = {
  className: 'btlr',
  cssName: 'border-top-left-radius',
  jsName: 'borderTopLeftRadius'
};
var _borderTopRightRadius = {
  className: 'btrr',
  cssName: 'border-top-right-radius',
  jsName: 'borderTopRightRadius'
};
var _borderBottomLeftRadius = {
  className: 'bblr',
  cssName: 'border-bottom-left-radius',
  jsName: 'borderBottomLeftRadius'
};
var _borderBottomRightRadius = {
  className: 'bbrr',
  cssName: 'border-bottom-right-radius',
  jsName: 'borderBottomRightRadius'
};

var propEnhancers = exports.propEnhancers = {
  borderBottomLeftRadius: function borderBottomLeftRadius(value) {
    return (0, _getCss2.default)(_borderBottomLeftRadius, value);
  },
  borderBottomRightRadius: function borderBottomRightRadius(value) {
    return (0, _getCss2.default)(_borderBottomRightRadius, value);
  },
  borderTopLeftRadius: function borderTopLeftRadius(value) {
    return (0, _getCss2.default)(_borderTopLeftRadius, value);
  },
  borderTopRightRadius: function borderTopRightRadius(value) {
    return (0, _getCss2.default)(_borderTopRightRadius, value);
  }
};
