// ui-box/lib/enhancers/dimensions.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.propEnhancers = exports.propValidators = exports.propAliases = exports.propTypes = undefined;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _getCss = require('../get-css');

var _getCss2 = _interopRequireDefault(_getCss);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var propTypes = exports.propTypes = {
  height: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  maxHeight: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  maxWidth: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  minHeight: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  minWidth: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  width: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number])
};

var propAliases = exports.propAliases = {};

var propValidators = exports.propValidators = {};

var _width = {
  className: 'w',
  cssName: 'width',
  jsName: 'width'
};
var _height = {
  className: 'h',
  cssName: 'height',
  jsName: 'height'
};
var _minWidth = {
  className: 'min-w',
  cssName: 'min-width',
  jsName: 'minWidth'
};
var _minHeight = {
  className: 'min-h',
  cssName: 'min-height',
  jsName: 'minHeight'
};
var _maxWidth = {
  className: 'max-w',
  cssName: 'max-width',
  jsName: 'maxWidth'
};
var _maxHeight = {
  className: 'max-h',
  cssName: 'max-height',
  jsName: 'maxHeight'
};

var propEnhancers = exports.propEnhancers = {
  height: function height(value) {
    return (0, _getCss2.default)(_height, value);
  },
  maxHeight: function maxHeight(value) {
    return (0, _getCss2.default)(_maxHeight, value);
  },
  maxWidth: function maxWidth(value) {
    return (0, _getCss2.default)(_maxWidth, value);
  },
  minHeight: function minHeight(value) {
    return (0, _getCss2.default)(_minHeight, value);
  },
  minWidth: function minWidth(value) {
    return (0, _getCss2.default)(_minWidth, value);
  },
  width: function width(value) {
    return (0, _getCss2.default)(_width, value);
  }
};
