// ui-box/lib/enhancers/interaction.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.propEnhancers = exports.propValidators = exports.propAliases = exports.propTypes = undefined;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _getCss = require('../get-css');

var _getCss2 = _interopRequireDefault(_getCss);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var propTypes = exports.propTypes = {
  cursor: _propTypes2.default.string,
  pointerEvents: _propTypes2.default.string,
  userSelect: _propTypes2.default.string,
  visibility: _propTypes2.default.string
};

var propAliases = exports.propAliases = {};

var propValidators = exports.propValidators = {};

var _cursor = {
  className: 'crsr',
  cssName: 'cursor',
  jsName: 'cursor'
};
var _userSelect = {
  className: 'usr-slct',
  cssName: 'user-select',
  jsName: 'userSelect',
  safeValue: true,
  isPrefixed: true
};
var _visibility = {
  className: 'vsblt',
  cssName: 'visibility',
  jsName: 'visibility',
  safeValue: true
};
var _pointerEvents = {
  className: 'ptr-evts',
  cssName: 'pointer-events',
  jsName: 'pointerEvents',
  safeValue: true
};

var propEnhancers = exports.propEnhancers = {
  cursor: function cursor(value) {
    return (0, _getCss2.default)(_cursor, value);
  },
  pointerEvents: function pointerEvents(value) {
    return (0, _getCss2.default)(_pointerEvents, value);
  },
  userSelect: function userSelect(value) {
    return (0, _getCss2.default)(_userSelect, value);
  },
  visibility: function visibility(value) {
    return (0, _getCss2.default)(_visibility, value);
  }
};
