// ui-box/lib/enhancers/list.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.propEnhancers = exports.propValidators = exports.propAliases = exports.propTypes = undefined;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _getCss = require('../get-css');

var _getCss2 = _interopRequireDefault(_getCss);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var propTypes = exports.propTypes = {
  listStyle: _propTypes2.default.string,
  listStyleType: _propTypes2.default.string,
  listStyleImage: _propTypes2.default.string,
  listStylePosition: _propTypes2.default.string
};

var propAliases = exports.propAliases = {};

var propValidators = exports.propValidators = {};

var _listStyle = {
  className: 'ls',
  cssName: 'list-style',
  jsName: 'listStyle',
  complexValue: true
};
var _listStyleType = {
  className: 'ls-typ',
  cssName: 'list-style-type',
  jsName: 'listStyleType'
};
var _listStyleImage = {
  className: 'ls-img',
  cssName: 'list-style-image',
  jsName: 'listStyleImage',
  complexValue: true
};
var _listStylePosition = {
  className: 'ls-pos',
  cssName: 'list-style-position',
  jsName: 'listStylePosition',
  safeValue: true
};

var propEnhancers = exports.propEnhancers = {
  listStyle: function listStyle(value) {
    return (0, _getCss2.default)(_listStyle, value);
  },
  listStyleType: function listStyleType(value) {
    return (0, _getCss2.default)(_listStyleType, value);
  },
  listStyleImage: function listStyleImage(value) {
    return (0, _getCss2.default)(_listStyleImage, value);
  },
  listStylePosition: function listStylePosition(value) {
    return (0, _getCss2.default)(_listStylePosition, value);
  }
};
