// ui-box/lib/enhancers/overflow.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.propEnhancers = exports.propValidators = exports.propAliases = exports.propTypes = undefined;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _getCss = require('../get-css');

var _getCss2 = _interopRequireDefault(_getCss);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var propTypes = exports.propTypes = {
  overflow: _propTypes2.default.string,
  overflowX: _propTypes2.default.string,
  overflowY: _propTypes2.default.string
};

var propAliases = exports.propAliases = {
  overflow: ['overflowX', 'overflowY']
};

var propValidators = exports.propValidators = {};

var _overflowY = {
  className: 'ovflw-y',
  cssName: 'overflow-y',
  jsName: 'overflowY',
  safeValue: true
};
var _overflowX = {
  className: 'ovflw-x',
  cssName: 'overflow-x',
  jsName: 'overflowX',
  safeValue: true
};

var propEnhancers = exports.propEnhancers = {
  overflowX: function overflowX(value) {
    return (0, _getCss2.default)(_overflowX, value);
  },
  overflowY: function overflowY(value) {
    return (0, _getCss2.default)(_overflowY, value);
  }
};
