// ui-box/lib/enhancers/spacing.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.propEnhancers = exports.propValidators = exports.propAliases = exports.propTypes = undefined;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _getCss = require('../get-css');

var _getCss2 = _interopRequireDefault(_getCss);

var _regex = require('../utils/regex');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var propTypes = exports.propTypes = {
  margin: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  marginBottom: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  marginLeft: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  marginRight: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  marginTop: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  marginX: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  marginY: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  padding: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  paddingBottom: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  paddingLeft: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  paddingRight: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  paddingTop: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  paddingX: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  paddingY: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number])
};

var propAliases = exports.propAliases = {
  margin: ['marginBottom', 'marginLeft', 'marginRight', 'marginTop'],
  marginX: ['marginLeft', 'marginRight'],
  marginY: ['marginBottom', 'marginTop'],
  padding: ['paddingBottom', 'paddingLeft', 'paddingRight', 'paddingTop'],
  paddingX: ['paddingLeft', 'paddingRight'],
  paddingY: ['paddingBottom', 'paddingTop']
};

var propValidators = exports.propValidators = {};

if (process.env.NODE_ENV !== 'production') {
  propValidators.margin = function (value) {
    if (_regex.spacesOutsideParentheses.test(value)) {
      return 'multiple values (\u201C' + value + '\u201D) aren\u055At supported with \u201Cmargin\u201D. Use \u201CmarginX\u201D, \u201CmarginY\u201D \u201CmarginBottom\u201D, \u201CmarginLeft\u201D, \u201CmarginRight\u201D and \u201CmarginTop\u201D instead.';
    }
  };
  propValidators.marginX = function (value) {
    if (_regex.spacesOutsideParentheses.test(value)) {
      return 'multiple values (\u201C' + value + '\u201D) aren\u055At supported with \u201CmarginX\u201D. Use \u201CmarginLeft\u201D and \u201CmarginRight\u201D instead.';
    }
  };
  propValidators.marginY = function (value) {
    if (_regex.spacesOutsideParentheses.test(value)) {
      return 'multiple values (\u201C' + value + '\u201D) aren\u055At supported with \u201CmarginY\u201D. Use \u201CmarginBottom\u201D and \u201CmarginTop\u201D instead.';
    }
  };
  propValidators.padding = function (value) {
    if (_regex.spacesOutsideParentheses.test(value)) {
      return 'multiple values (\u201C' + value + '\u201D) aren\u055At supported with \u201Cpadding\u201D. Use \u201CpaddingX\u201D, \u201CpaddingY\u201D \u201CpaddingBottom\u201D, \u201CpaddingLeft\u201D, \u201CpaddingRight\u201D and \u201CpaddingTop\u201D instead.';
    }
  };
  propValidators.paddingX = function (value) {
    if (_regex.spacesOutsideParentheses.test(value)) {
      return 'multiple values (\u201C' + value + '\u201D) aren\u055At supported with \u201CpaddingX\u201D. Use \u201CpaddingLeft\u201D and \u201CpaddingRight\u201D instead.';
    }
  };
  propValidators.paddingY = function (value) {
    if (_regex.spacesOutsideParentheses.test(value)) {
      return 'multiple values (\u201C' + value + '\u201D) aren\u055At supported with \u201CpaddingY\u201D. Use \u201CpaddingBottom\u201D and \u201CpaddingTop\u201D instead.';
    }
  };
}

var _marginTop = {
  className: 'mt',
  cssName: 'margin-top',
  jsName: 'marginTop'
};
var _marginRight = {
  className: 'mr',
  cssName: 'margin-right',
  jsName: 'marginRight'
};
var _marginBottom = {
  className: 'mb',
  cssName: 'margin-bottom',
  jsName: 'marginBottom'
};
var _marginLeft = {
  className: 'ml',
  cssName: 'margin-left',
  jsName: 'marginLeft'
};
var _paddingTop = {
  className: 'pt',
  cssName: 'padding-top',
  jsName: 'paddingTop'
};
var _paddingRight = {
  className: 'pr',
  cssName: 'padding-right',
  jsName: 'paddingRight'
};
var _paddingBottom = {
  className: 'pb',
  cssName: 'padding-bottom',
  jsName: 'paddingBottom'
};
var _paddingLeft = {
  className: 'pl',
  cssName: 'padding-left',
  jsName: 'paddingLeft'
};

var propEnhancers = exports.propEnhancers = {
  marginBottom: function marginBottom(value) {
    return (0, _getCss2.default)(_marginBottom, value);
  },
  marginLeft: function marginLeft(value) {
    return (0, _getCss2.default)(_marginLeft, value);
  },
  marginRight: function marginRight(value) {
    return (0, _getCss2.default)(_marginRight, value);
  },
  marginTop: function marginTop(value) {
    return (0, _getCss2.default)(_marginTop, value);
  },
  paddingBottom: function paddingBottom(value) {
    return (0, _getCss2.default)(_paddingBottom, value);
  },
  paddingLeft: function paddingLeft(value) {
    return (0, _getCss2.default)(_paddingLeft, value);
  },
  paddingRight: function paddingRight(value) {
    return (0, _getCss2.default)(_paddingRight, value);
  },
  paddingTop: function paddingTop(value) {
    return (0, _getCss2.default)(_paddingTop, value);
  }
};
