// ui-box/lib/enhancers/transition.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.propEnhancers = exports.propValidators = exports.propAliases = exports.propTypes = undefined;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _getCss = require('../get-css');

var _getCss2 = _interopRequireDefault(_getCss);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var propTypes = exports.propTypes = {
  transition: _propTypes2.default.string,
  transitionDelay: _propTypes2.default.string,
  transitionDuration: _propTypes2.default.string,
  transitionProperty: _propTypes2.default.string,
  transitionTimingFunction: _propTypes2.default.string
};

var propAliases = exports.propAliases = {};

var propValidators = exports.propValidators = {};

var _transition = {
  className: 'tstn',
  cssName: 'transition',
  jsName: 'transition',
  complexValue: true
};
var _transitionDelay = {
  className: 'tstn-dly',
  cssName: 'transition-delay',
  jsName: 'transitionDelay',
  complexValue: true
};
var _transitionDuration = {
  className: 'tstn-drn',
  cssName: 'transition-duration',
  jsName: 'transitionDuration',
  complexValue: true
};
var _transitionProperty = {
  className: 'tstn-pty',
  cssName: 'transition-property',
  jsName: 'transitionProperty',
  complexValue: true
};
var _transitionTimingFunction = {
  className: 'tstn-tf',
  cssName: 'transition-timing-function',
  jsName: 'transitionTimingFunction',
  complexValue: true
};

var propEnhancers = exports.propEnhancers = {
  transition: function transition(value) {
    return (0, _getCss2.default)(_transition, value);
  },
  transitionDelay: function transitionDelay(value) {
    return (0, _getCss2.default)(_transitionDelay, value);
  },
  transitionDuration: function transitionDuration(value) {
    return (0, _getCss2.default)(_transitionDuration, value);
  },
  transitionProperty: function transitionProperty(value) {
    return (0, _getCss2.default)(_transitionProperty, value);
  },
  transitionTimingFunction: function transitionTimingFunction(value) {
    return (0, _getCss2.default)(_transitionTimingFunction, value);
  }
};
