// evergreen-ui/esm/table/src/TextTableHeaderCell.js
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread";
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { Text } from '../../typography';
import TableHeaderCell from './TableHeaderCell';

var TextTableHeaderCell =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(TextTableHeaderCell, _PureComponent);

  function TextTableHeaderCell() {
    _classCallCheck(this, TextTableHeaderCell);

    return _possibleConstructorReturn(this, _getPrototypeOf(TextTableHeaderCell).apply(this, arguments));
  }

  _createClass(TextTableHeaderCell, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          children = _this$props.children,
          textProps = _this$props.textProps,
          isSortable = _this$props.isSortable,
          sortOrder = _this$props.sortOrder,
          props = _objectWithoutProperties(_this$props, ["children", "textProps", "isSortable", "sortOrder"]);

      return React.createElement(TableHeaderCell, props, React.createElement(Text, _extends({
        fontWeight: 500,
        size: 300,
        flex: "1"
      }, textProps), children, ' '));
    }
  }]);

  return TextTableHeaderCell;
}(PureComponent);

TextTableHeaderCell.displayName = "TextTableHeaderCell";

_defineProperty(TextTableHeaderCell, "propTypes", _objectSpread({}, TableHeaderCell.propTypes, {
  /**
   * Pass additional props to the Text component.
   */
  textProps: PropTypes.objectOf(PropTypes.object)
}));

export { TextTableHeaderCell as default };
