// evergreen-ui/commonjs/menu/src/MenuGroup.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _layers = require("../../layers");

var _typography = require("../../typography");

var MenuGroup =
/*#__PURE__*/
function (_React$PureComponent) {
  (0, _inherits2.default)(MenuGroup, _React$PureComponent);

  function MenuGroup() {
    (0, _classCallCheck2.default)(this, MenuGroup);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(MenuGroup).apply(this, arguments));
  }

  (0, _createClass2.default)(MenuGroup, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          title = _this$props.title,
          children = _this$props.children;
      return _react.default.createElement(_layers.Pane, {
        paddingY: 8
      }, title && _react.default.createElement(_typography.Heading, {
        size: 100,
        marginX: 16,
        marginY: 8
      }, title), children);
    }
  }]);
  return MenuGroup;
}(_react.default.PureComponent);

exports.default = MenuGroup;
MenuGroup.displayName = "MenuGroup";
(0, _defineProperty2.default)(MenuGroup, "propTypes", {
  /**
   * Title of the menu group.
   */
  title: _propTypes.default.node,

  /**
   * The children of the menu group.
   */
  children: _propTypes.default.node
});
