// evergreen-ui/commonjs/theme/src/default-theme/component-specific/getLinkClassName.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _tinycolor = _interopRequireDefault(require("tinycolor2"));

var _themer = require("../../../../themer");

var _memoizeClassName = _interopRequireDefault(require("../utils/memoizeClassName"));

var _palette = _interopRequireDefault(require("../foundational-styles/palette"));

/**
 * The link appearance unlike the Button is based on the color property.
 * Currently the Link does not support the Intent or the appearance interface.
 * @param {string} color
 * @return {Object} appearance of the link.
 */
var getLinkAppearance = function getLinkAppearance(color) {
  switch (color) {
    case 'neutral':
      return _themer.Themer.createLinkAppearance({
        base: {
          color: _palette.default.neutral.base
        },
        hover: {
          color: (0, _tinycolor.default)(_palette.default.neutral.base).lighten(10).toString()
        },
        active: {
          color: (0, _tinycolor.default)(_palette.default.neutral.base).darken(10).toString()
        },
        focus: {
          boxShadow: "0 0 0 2px ".concat((0, _tinycolor.default)(_palette.default.neutral.base).setAlpha(0.4).toString())
        }
      });

    case 'green':
      return _themer.Themer.createLinkAppearance({
        base: {
          color: _palette.default.green.base
        },
        hover: {
          color: (0, _tinycolor.default)(_palette.default.green.base).lighten(10).toString()
        },
        active: {
          color: (0, _tinycolor.default)(_palette.default.green.base).darken(10).toString()
        },
        focus: {
          boxShadow: "0 0 0 2px ".concat((0, _tinycolor.default)(_palette.default.green.base).setAlpha(0.4).toString())
        }
      });

    case 'default':
    case 'blue':
    default:
      return _themer.Themer.createLinkAppearance({
        base: {
          color: _palette.default.blue.base
        },
        hover: {
          color: (0, _tinycolor.default)(_palette.default.blue.base).lighten(10).toString()
        },
        active: {
          color: (0, _tinycolor.default)(_palette.default.blue.base).darken(10).toString()
        },
        focus: {
          boxShadow: "0 0 0 2px ".concat((0, _tinycolor.default)(_palette.default.blue.base).setAlpha(0.4).toString())
        }
      });
  }
};
/**
 * Get the className of a `Link` component.
 * @param {string} color
 * @return {string} the appearance class name.
 */


var _default = (0, _memoizeClassName.default)(getLinkAppearance);

exports.default = _default;
