// evergreen-ui/commonjs/theme/src/default-theme/component-specific/getSwitchClassName.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _themer = require("../../../../themer");

var _memoizeClassName = _interopRequireDefault(require("../utils/memoizeClassName"));

var _scales = _interopRequireDefault(require("../foundational-styles/scales"));

var defaultAppearance = _themer.Themer.createSwitchAppearance({
  base: {
    transition: 'all 120ms ease-in-out',
    cursor: 'pointer',
    color: 'white',
    backgroundColor: _scales.default.neutral.N5A,
    borderRadius: 9999
  },
  disabled: {
    opacity: 0.5,
    backgroundImage: 'none'
  },
  hover: {
    backgroundColor: _scales.default.neutral.N5A
  },
  active: {
    backgroundColor: _scales.default.neutral.N6A
  },
  focus: {
    boxShadow: "0 0 0 3px ".concat(_scales.default.blue.B6A)
  },
  checked: {
    backgroundColor: _scales.default.blue.B8,
    color: 'white'
  },
  checkedHover: {
    backgroundColor: _scales.default.blue.B8,
    color: 'white'
  },
  checkedActive: {
    backgroundColor: _scales.default.blue.B9,
    color: 'white'
  },
  checkedDisabled: {}
});
/**
 * Get the className of a `Switch`.
 * @param {string} appearance
 * @return {Object} the appearance object.
 */


var getSwitchAppearance = function getSwitchAppearance() {
  return defaultAppearance;
};
/**
 * Get the className of a `Switch`.
 * @param {string} appearance
 * @return {string} the appearance class name.
 */


var _default = (0, _memoizeClassName.default)(getSwitchAppearance);

exports.default = _default;
