// evergreen-ui/commonjs/theme/src/default-theme/component-specific/getSegmentedControlRadioClassName.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _themer = require("../../../../themer");

var _memoizeClassName = _interopRequireDefault(require("../utils/memoizeClassName"));

var _shared = require("../shared");

var defaultAppearance = _themer.Themer.createSegmentedControlRadioAppearance({
  base: _shared.defaultControlStyles.base,
  disabled: _shared.defaultControlStyles.disabled,
  hover: _shared.defaultControlStyles.hover,
  active: _shared.defaultControlStyles.active,
  focus: _shared.defaultControlStyles.focus
});
/**
 * Get the appearanece of a `SegmentedControlRadio`.
 * @param {string} appearance
 * @return {string} the appearance object.
 */


var getSegmentedControlRadioAppearance = function getSegmentedControlRadioAppearance() {
  return defaultAppearance;
};
/**
 * Get the className of a `SegmentedControlRadio`.
 * @param {string} appearance
 * @return {string} the appearance class name.
 */


var _default = (0, _memoizeClassName.default)(getSegmentedControlRadioAppearance);

exports.default = _default;
