// evergreen-ui/commonjs/theme/src/default-theme/component-specific/getTextDropdownButtonClassName.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _themer = require("../../../../themer");

var _memoizeClassName = _interopRequireDefault(require("../utils/memoizeClassName"));

var _scales = _interopRequireDefault(require("../foundational-styles/scales"));

var defaultAppearance = _themer.Themer.createTextDropdownButtonAppearance({
  base: {
    borderRadius: 3
  },
  hover: {},
  focus: {
    boxShadow: "0 0 0 3px ".concat(_scales.default.blue.B5A)
  },
  active: {},
  disabled: {
    opacity: 0.5
  }
});
/**
 * Get the appearance of a `TextDropdownButton`.
 */


var getTextDropdownButtonAppearance = function getTextDropdownButtonAppearance() {
  return defaultAppearance;
};
/**
 * Get the className of a `TextDropdownButton`.
 */


var _default = (0, _memoizeClassName.default)(getTextDropdownButtonAppearance);

exports.default = _default;
