// react-custom-validation/.next/dist/pages/example/1.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('/home/marcelka/projects/validation/node_modules/next/node_modules/babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _inherits2 = require('/home/marcelka/projects/validation/node_modules/next/node_modules/babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _defineProperty2 = require('/home/marcelka/projects/validation/node_modules/next/node_modules/babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _getPrototypeOf = require('/home/marcelka/projects/validation/node_modules/next/node_modules/babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('/home/marcelka/projects/validation/node_modules/next/node_modules/babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('/home/marcelka/projects/validation/node_modules/next/node_modules/babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('/home/marcelka/projects/validation/node_modules/next/node_modules/babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _react = require('/home/marcelka/projects/validation/node_modules/react/react.js');

var _react2 = _interopRequireDefault(_react);

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

var _immutabilityHelper = require('immutability-helper');

var _immutabilityHelper2 = _interopRequireDefault(_immutabilityHelper);

var _validator = require('validator');

var _validator2 = _interopRequireDefault(_validator);

var _style = require('./style1');

var _style2 = _interopRequireDefault(_style);

var _lib = require('../../lib');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var App = function (_React$Component) {
  (0, _inherits3.default)(App, _React$Component);

  function App() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, App);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = App.__proto__ || (0, _getPrototypeOf2.default)(App)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      fields: {
        email: '',
        password: '',
        rePassword: ''
      }
    }, _this.fieldChange = function (field, value) {
      _this.setState((0, _immutabilityHelper2.default)(_this.state, { fields: (0, _defineProperty3.default)({}, field, { $set: value }) }));
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(App, [{
    key: 'render',
    value: function render() {
      return _react2.default.createElement(Form, {
        fields: this.state.fields,
        onChange: this.fieldChange,
        onValid: function onValid() {
          return alert('Submitting...');
        } // eslint-disable-line no-alert
        , onInvalid: function onInvalid() {
          return alert('Error!');
        } // eslint-disable-line no-alert
      });
    }
  }]);
  return App;
}(_react2.default.Component);

var isEmail = function isEmail(email) {
  return _validator2.default.isEmail(email) ? null : 'This is not a valid email.';
};

var isUnique = function isUnique(email) {
  return _bluebird2.default.delay(1000).then(function () {
    return email.includes('used') ? 'This email is already used.' : null;
  });
};

var minLength = function minLength(password, length) {
  return password.length >= length ? null : 'Password is too short.';
};

var areSame = function areSame(password, rePassword) {
  return password === rePassword ? null : 'Passwords do not match.';
};

function validationConfig(props) {
  var _props$fields = props.fields,
      email = _props$fields.email,
      password = _props$fields.password,
      rePassword = _props$fields.rePassword;


  return {
    fields: ['email', 'password', 'rePassword'],

    validations: {
      email: [[isEmail, email], [isUnique, email]],
      password: [[minLength, password, 6]],
      rePassword: {
        rules: [[areSame, password, rePassword]],
        fields: ['password', 'rePassword']
      }
    }
  };
}

var Form = function (_React$Component2) {
  (0, _inherits3.default)(Form, _React$Component2);

  function Form() {
    (0, _classCallCheck3.default)(this, Form);
    return (0, _possibleConstructorReturn3.default)(this, (Form.__proto__ || (0, _getPrototypeOf2.default)(Form)).apply(this, arguments));
  }

  (0, _createClass3.default)(Form, [{
    key: 'render',
    value: function render() {
      var _this3 = this;

      var _props = this.props,
          fields = _props.fields,
          onChange = _props.onChange,
          onValid = _props.onValid,
          onInvalid = _props.onInvalid,
          $field = _props.$field,
          $validation = _props.$validation;

      return _react2.default.createElement(
        'form',
        { className: _style2.default },
        _react2.default.createElement(
          'h1',
          null,
          'Sign up'
        ),
        _react2.default.createElement(
          'label',
          null,
          'Email'
        ),
        $validation.email.show && _react2.default.createElement(
          'span',
          null,
          $validation.email.error.reason
        ),
        _react2.default.createElement('input', (0, _extends3.default)({ type: 'text', value: fields.email
        }, $field('email', function (e) {
          return onChange('email', e.target.value);
        }))),
        _react2.default.createElement(
          'label',
          null,
          'Password'
        ),
        $validation.password.show && _react2.default.createElement(
          'span',
          null,
          $validation.password.error.reason
        ),
        _react2.default.createElement('input', (0, _extends3.default)({ type: 'password', value: fields.password
        }, $field('password', function (e) {
          return onChange('password', e.target.value);
        }))),
        _react2.default.createElement(
          'label',
          null,
          'Repeat password'
        ),
        $validation.rePassword.show && _react2.default.createElement(
          'span',
          null,
          $validation.rePassword.error.reason
        ),
        _react2.default.createElement('input', (0, _extends3.default)({ type: 'password', value: fields.rePassword
        }, $field('rePassword', function (e) {
          return onChange('rePassword', e.target.value);
        }))),
        _react2.default.createElement(
          'button',
          { onClick: function onClick(e) {
              e.preventDefault();
              _this3.props.$submit(onValid, onInvalid);
            } },
          'Sign up'
        )
      );
    }
  }]);
  return Form;
}(_react2.default.Component);

Form = (0, _lib.validated)(validationConfig)(Form);

exports.default = function () {
  return _react2.default.createElement(App, null);
};
    if (module.hot) {
      module.hot.accept()

      var Component = module.exports.default || module.exports
      Component.__route = "/example/1"

      if (module.hot.status() !== 'idle') {
        var components = next.router.components
        for (var r in components) {
          if (!components.hasOwnProperty(r)) continue

          if (components[r].Component.__route === "/example/1") {
            next.router.update(r, Component)
          }
        }
      }
    }
