// react-custom-validation/.next/dist/pages/example/style1.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _css = require('/home/marcelka/projects/validation/node_modules/next/dist/lib/css.js');

var _css2 = _interopRequireDefault(_css);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = _css2.default.merge((0, _css2.default)({
  width: '95%',
  maxWidth: '20rem',
  margin: '0 auto',
  fontFamily: 'Helvetica, Arial, sans-serif',
  boxSizing: 'border-box',
  textAlign: 'center'
}), _css2.default.$(' h1', {
  marginTop: '0',
  color: '#444'
}), _css2.default.$(' input', {
  display: 'block',
  height: '2rem',
  width: '100%',
  padding: '0 .5rem',
  marginBottom: '1rem',
  borderRadius: '.25rem',
  border: '1px solid #ddd',
  boxSizing: 'border-box'
}), _css2.default.$(' input:focus', {
  outline: 'none',
  borderColor: '#aaa'
}), _css2.default.$(' button', {
  height: '2rem',
  width: '5rem',
  textTransform: 'uppercase',
  borderRadius: '.25rem',
  border: '0',
  backgroundColor: '#dc0067',
  color: 'white'
}), _css2.default.$(' label', {
  display: 'block',
  color: '#444',
  float: 'left',
  marginBottom: '.5rem',
  fontSize: '0.9rem',
  paddingLeft: '.4rem'
}), _css2.default.$(' span', {
  color: '#d22424',
  float: 'right',
  display: 'block',
  fontSize: '0.9rem',
  paddingRight: '.4rem'
}));
    if (module.hot) {
      module.hot.accept()

      var Component = module.exports.default || module.exports
      Component.__route = "/example/style1"

      if (module.hot.status() !== 'idle') {
        var components = next.router.components
        for (var r in components) {
          if (!components.hasOwnProperty(r)) continue

          if (components[r].Component.__route === "/example/style1") {
            next.router.update(r, Component)
          }
        }
      }
    }
